src/dmd/root/file.d

```diff
@@ -22,6 +23,376 @@ import dmd.root.filename;
 import dmd.root.rmem;
 import dmd.root.string;
 
+/**
+Encapsulated management of a memory-mapped file.
+
+The `Datum` type encodes the mapping mode: Use `ubyte` for read/write mapping
+and `const ubyte` for read-only mapping.
+*/
+struct FileMapping(Datum)
+{
+    static assert(Datum.sizeof == 1, "Not tested with other data sizes yet.");
+
+    version(Posix) enum invalidHandle = -1;
+    else version(Windows) enum invalidHandle = INVALID_HANDLE_VALUE;
+
+    // state {
+    /// Handle of underlying file
+    private auto handle = invalidHandle;
+    /// File mapping object needed on Windows
+    version(Windows) private HANDLE fileMappingObject = invalidHandle;
+    /// Memory-mapped array
+    private Datum[] data;
+    /// Name of underlying file, zero-terminated
+    private const(char)* name;
+    // state }
+
+    /**
+    Open `filename` and map in memory. If `Datum` is `const`, opens for
+    read-only and maps the content in memory; no error is issued if the file
+    does not exist. This makes it easy to treat a non-existing file as empty.
+
+    If `Datum` is mutable, opens for read/write (creates file if it does not
+    exist) and fails fatally on any error.
+
+    Due to quirks in `mmap`, if the file is empty, `handle` is valid but `data`
+    is `null`.
+    */
+    this(const char* filename)
+    {
+        version (Posix)
+        {
+            import core.sys.posix.sys.mman;
+            import core.sys.posix.fcntl;
+
+            handle = .open(filename, is(Datum == const) ? O_RDONLY : (O_CREAT | O_RDWR),
+                S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
+
+            if (handle == -1)
+            {
+                static if (is(Datum == const))
+                {
+                    // No error, nonexisting file in read mode behaves like an empty file.
+                    return;
+                }
+                else
+                {
+                    fprintf(stderr, "open(\"%s\") failed: %s\n", filename, strerror(errno));
+                    assert(0);
+                }
+            }
+
+            const size = File.size(handle);
```

Comment:
I note the subsequent `cast(size_t)` being applied to `size`. Trouble comes when `size_t` is a 32 bit value. Files can be larger than 32 bits, and truncating to 32 bits is going to cause very unhappy errors.

At minimum, add an assert() to assure that the cast will not truncate the file.

Reply:
Ah, I see you already do this.

Reply:
added a comment for good measure